src/protocols/screencopy.rs

```diff
@@ -0,0 +1,279 @@
+            smithay::reexports::wayland_protocols_wlr::screencopy::v1::server::zwlr_screencopy_frame_v1::ZwlrScreencopyFrameV1: $crate::protocols::screencopy::ScreencopyFrameState
+        ] => $crate::protocols::screencopy::ScreencopyManagerState);
+    };
+}
+
+#[derive(Clone)]
+pub struct ScreencopyFrameState {
+    pub output: Output,
+    pub rect: Rectangle<i32, Physical>,
+    pub overlay_cursor: bool,
+}
+
+impl<D> Dispatch<ZwlrScreencopyFrameV1, ScreencopyFrameState, D> for ScreencopyManagerState
+where
+    D: Dispatch<ZwlrScreencopyFrameV1, ScreencopyFrameState>,
+    D: ScreencopyHandler,
+    D: 'static,
+{
+    fn request(
+        state: &mut D,
+        _client: &Client,
+        frame: &ZwlrScreencopyFrameV1,
+        request: zwlr_screencopy_frame_v1::Request,
+        data: &ScreencopyFrameState,
+        _display: &DisplayHandle,
+        _data_init: &mut DataInit<'_, D>,
+    ) {
+        let (buffer, send_damage) = match request {
+            zwlr_screencopy_frame_v1::Request::Copy { buffer } => (buffer, false),
+            zwlr_screencopy_frame_v1::Request::CopyWithDamage { buffer } => (buffer, true),
+            zwlr_screencopy_frame_v1::Request::Destroy => return,
+            _ => unreachable!(),
+        };
+
+        state.frame(Screencopy {
+            send_damage,
+            buffer,
+            frame: frame.clone(),
+            frame_state: data.clone(),
+            submitted: false,
+        });
+    }
+}
+
+/// Screencopy frame.
+pub struct Screencopy {
+    frame_state: ScreencopyFrameState,
+    frame: ZwlrScreencopyFrameV1,
+    send_damage: bool,
+    buffer: WlBuffer,
+    submitted: bool,
+}
+
+impl Drop for Screencopy {
+    fn drop(&mut self) {
+        if !self.submitted {
+            self.frame.failed();
+        }
+    }
+}
+
+impl Screencopy {
+    /// Get the target buffer to copy to.
+    pub fn buffer(&self) -> &WlBuffer {
+        &self.buffer
+    }
+
+    /// Get the region which should be copied.
+    pub fn region(&self) -> Rectangle<i32, Physical> {
+        self.frame_state.rect
+    }
+
+    pub fn output(&self) -> &Output {
+        &self.frame_state.output
+    }
+
+    pub fn overlay_cursor(&self) -> bool {
+        self.frame_state.overlay_cursor
+    }
+
+    /// Mark damaged regions of the screencopy buffer.
+    pub fn damage(&mut self, damage: &[Rectangle<i32, Physical>]) {
+        if !self.send_damage {
+            return;
+        }
+
+        for Rectangle { loc, size } in damage {
+            self.frame
+                .damage(loc.x as u32, loc.y as u32, size.w as u32, size.h as u32);
+        }
+    }
+
+    /// Submit the copied content.
+    pub fn submit(mut self) {
+        // Notify client that buffer is ordinary.
+        self.frame.flags(Flags::empty());
+
+        // Notify client about successful copy.
+        let now = UNIX_EPOCH.elapsed().unwrap();
```

Comment:
We can send the real presentation time here.

Reply:
Isn't that what this does? i.e. "time elapsed since the UNIX epoch"

Reply:
That's the time when the function is called, however we have the precise time from the GPU when the frame is shown on screen.

Actually, come to think of it, this would require sending a screencopy on vblank rather than on redraw (since the presentation timestamp is only available on vblank), so this is fine for now.

Reply:
ah. didn't catch what "presentation time" means. when i first glanced at this code in catacomb, i thought it was sending just unix epoch, and read your suggestion as just "send the real time".